### PR TITLE
Improve metrics collecting to differentiate crashes and user_errors

### PR DESCRIPTION
**Pull request by mfurtak**
And track metrics by tool versions

### CODE REVIEW

File: app/controllers/bacons_controller.rb [hunk 1]
```diff
@@ -11,19 +12,28 @@ def did_launch
     # Store the number of runs per action
     now = Time.now.to_date
     launches.each do |action, count|
-      entry = Bacon.where(action_name: action, launch_date: now).take
-      entry ||= Bacon.create!(action_name: action, launch_date: now, launches: 0, number_errors: 0)
-
-      entry.launches += count
+      entry = Bacon.find_or_create_by(action_name: action, launch_date: now, tool_version: versions[action])
+      entry.increment(:launches, count)
```
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb:16]: So much magic I didn't know of

File: app/controllers/bacons_controller.rb [hunk 2]
```diff
@@ -11,19 +12,28 @@ def did_launch
...
       entry.save
     end
 
-    # Store the error information
     error_name = params[:error]
```
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb]: What's an example value of this? What's the `error_name`?

Comment (KrauseFx) [inline app/controllers/bacons_controller.rb]: The name of the action?

File: app/controllers/bacons_controller.rb [hunk 3]
```diff
@@ -56,6 +66,10 @@ def stats
     ]
   end
 
+  def versions
+    @versions ||= JSON.parse(params[:versions]) rescue Hash.new('unknown')
```
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb]:
``` ruby
Hash.new('unknown')
```

returns
`{}` 
for me. Is that supposed to be like this?

File: app/controllers/bacons_controller.rb [hunk 4]
```diff
@@ -98,7 +112,7 @@ def graphs
   private
     def authenticate
       authenticate_or_request_with_http_basic do |username, password|
-        username == "admin" && password == ENV["FL_PASSWORD"]
+        username == "admin" && password == ENV["  "]
```
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb]: Commited by mistaked?

File: config/database.yml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 development:
-  adapter: postgresql
-  database: enhancer
+  adapter: sqlite3
```
Comment (KrauseFx) [inline config/database.yml]: Commited by mistake?

File: db/schema.rb [hunk 6]
```diff
@@ -11,15 +11,17 @@
...
-    t.datetime "created_at",    null: false
-    t.datetime "updated_at",    null: false
+    t.string   "action_name",                        null: false
+    t.string   "launch_date",                        null: false
+    t.integer  "launches",       default: 0,         null: false
+    t.integer  "number_errors",  default: 0,         null: false
+    t.datetime "created_at",                         null: false
+    t.datetime "updated_at",                         null: false
+    t.string   "tool_version",   default: "unknown", null: false
+    t.integer  "number_crashes", default: 0,         null: false
```
Comment (i2amsam) [inline db/schema.rb]: Separate the migration into a different PR you can ship first so you don't get into a state with new code running against the old schema.

File: app/controllers/bacons_controller.rb (hunk 4 above)
Comment (i2amsam) [inline app/controllers/bacons_controller.rb]: Add the password back

File: app/controllers/bacons_controller.rb (hunk 3 above)
Comment (mfurtak) [inline app/controllers/bacons_controller.rb]: Yep, this returns an empty Hash as the value of the constructor, but then the instance returns `'unknown'` instead of `nil` if you ask for a key that has no value

```
$ irb
>> Hash.new('unknown')['blah']
=> "unknown"
```

File: app/controllers/bacons_controller.rb [hunk 7]
```diff
@@ -11,19 +12,28 @@ def did_launch
...
+        b.increment(:number_errors)
+        b.save
+      end
+    end
+
+    crash_name = params[:crash]
+    # Store the crash information
+    if crash_name
+      crash_version = versions[crash_name]
+      Bacon.find_by(action_name: crash_name, launch_date: now, tool_version: crash_version).try do |b|
```
Comment (i2amsam) [inline app/controllers/bacons_controller.rb]: This could be DRYed

```
update_bacon_for(action, tool) do |bacon|
  bacon.increment(:the_item)
  bacon.save!
end
```

File: app/controllers/bacons_controller.rb (hunk 2 above)
Comment (mfurtak) [inline app/controllers/bacons_controller.rb]: Yes, the name of the action or tool. This has not changed, right?

File: db/schema.rb (hunk 6 above)
Comment (mfurtak) [inline db/schema.rb]: Let's talk about how that would happen.

File: app/controllers/bacons_controller.rb (hunk 1 above)
Comment (mfurtak) [inline app/controllers/bacons_controller.rb:16]: Rails is _so_ ✨ magic✨

File: app/controllers/bacons_controller.rb (hunk 3 above)
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb]: Oh, wow, I didn't know that, thanks for clarifying! 👍

Comment (mfurtak) [inline app/controllers/bacons_controller.rb]: We'll be changing it because it's wrong in another case, however! :smile:

**Comment by KrauseFx** [comment]
Oh you squashed instead of adding a new commit 😢

File: app/controllers/bacons_controller.rb [hunk 8]
```diff
@@ -11,26 +12,35 @@ def did_launch
     # Store the number of runs per action
     now = Time.now.to_date
     launches.each do |action, count|
-      entry = Bacon.where(action_name: action, launch_date: now).take
-      entry ||= Bacon.create!(action_name: action, launch_date: now, launches: 0, number_errors: 0)
-
-      entry.launches += count
+      entry = Bacon.find_or_create_by(action_name: action, launch_date: now, tool_version: tool_version(action))
```
Comment (KrauseFx) [inline app/controllers/bacons_controller.rb:15]: Did I understand correct that we could have the following in our database:

action, date, version, launches
`deliver`, `today`, `1.0`, `2`
`deliver`, `today`, `1.1`, `4`

So we have 2 rows if the version number is not the same, right?

Comment (KrauseFx) [inline app/controllers/bacons_controller.rb:15]: So those 3 columns are the primary key

**Comment by KrauseFx** [comment]
Looks great, just added one question

**Comment by mfurtak** [comment]
Yeah, sorry for the squash. 😞  It was necessary to deal with moving the migration to another PR.

File: app/controllers/bacons_controller.rb (hunk 8 above)
Comment (mfurtak) [inline app/controllers/bacons_controller.rb:15]: Correct, `action_name`, `launch_date`, and `version` are now the composite primary key that we look up when updating stats.

Comment (mfurtak) [inline app/controllers/bacons_controller.rb:15]: So as your example suggests, we can get info from different versions of the same tool on the same day, and we'll be able to compare their success/error rates 👍
